src/blk/kernel/KernelDevice.cc

```diff
@@ -571,10 +571,19 @@ void KernelDevice::_discard_update_threads()
 void KernelDevice::_discard_stop()
 {
   dout(10) << __func__ << dendl;
-
-  discard_stop = true;
-  _discard_update_threads();
```

Comment:
IMO it would be better to update _discard_update_threads() in a way to perform threads shutdown (both on full device shutdown and partial threads terminatio ) properly instead of having custom code here. So _discard_update_threads() has to "join" thrueads to be terminated rather than detaching from them.

The rationale for the is still existing potential race condition when some threads have been detached due to thread count decrease and then full device shutdown is occurring - those threads will escape the new shutdown scenario this way and might still cause mutex access issue(s). 
Not to mention that having uniform thread shutdown approach is more elegant.